### ISSUE TITLE
No new versions?

### ISSUE THREAD

**Issue by vosdev** [canonical/lxd-ui#945]
Looking at the previous releases we had a new version roughly every 1 or 2 weeks with +- 10 commits per version. Now it's been 4 months and there have been 233 commits since 0.12.

https://github.com/canonical/lxd-ui/compare/0.12...main

I'm curious why the huge change in release cycle?

There have been several features added since 0.12 that I would love to see in my LXD UI that I now have to resort back to CLI for.

**Comment by edlerd** [comment]
Thank you for the request. Please, a bit of patience :)

Our current timeline for releases is 5.0.4 next, then 6.2 and then 5.21.3. The latest of the three is targetted for the end of November. 6.2 will include all the new features. 5.0.4 and 5.21.3 will only contain fixes. I think only 5.21.3 will contain fixes from the UI.

I just created a new 0.13 tag for the `latest/candidate` channel. It will be available shortly and might be the one for 6.2. But there might as well be a newer tag at the time closer to 6.2 release date, so we might have 0.14 shipping with 6.2.

You can always try the `latest/edge` channel, it will include all features that landed in the main branch. But edge should be considered unstable and not be used in a production environment. Also, there is no way back from edge to a stable channel.

**Comment by vosdev** [comment]
It was just out of curiousity. I was used to monthly releases of LXD before the Canonical takeover and now also seeing the change in releases for lxd-ui I just wanted to ask :)

I absolutely love the UI so I would love to test out the new features.

I am running latest/stable on all my LXD instances. How does latest/edge work? Does that snap get rebuild every day with the newest pulls from main branch?

**Comment by edlerd** [comment]
Nothing in particular changed the LXD release cadence. It is good feedback that you expect more frequent releases, I'll check if we can get back there.

> I am running latest/stable on all my LXD instances. How does latest/edge work? Does that snap get rebuild every day with the newest pulls from main branch?

Yes, exactly. It doesn't get triggered by UI merges, but only by LXD merges to the main branch. The edge build pulls from the UI main branch, so it usually contains all UI changes. While there is a test pipeline involved before any merge, the edge channel is not considered unstable. It can and has in the past contained changes that can mess things up. The version in the stable channels are tested more thoroughly.

Long story short, we now have all the UI changes in the stable/candidate and all fixes in the 5.21/candidate channels. We are currently doing some more testing and both might be released soonish. You could try the candidate channel and give us feedback on any issues found.

**Comment by vosdev** [comment]
Hey edlerd,


I am very happy with some of the changes in 0.13 :)

- The ability to export/import an instance. 
- Migration button
- the disappearance of the Edit instance/configuration button
- More options in the configuration view of an instance
- Smaller UI. I no longer have to zoom out to 80% to get a whole view without scrollbars!
- Love the new icons. They're a better fit imo.

**The ability to export/import an instance.** 
i didn't expect it to immediately download the instance as a file to my laptop. Instead, I wanted to select a network filesystem on my LXD host. Maybe that is something that can be added in the future?

**Migration**
Great feature! Had to resort back to the cli two weeks ago to do this. Glad it's in the ui now, but I can only do one at a time. Can I request a feature to be able to select multiple VMs in the Instance overview so that we can migrate multiple instances at a time? I don't really want to have to do so many clicks for 50 separate instances.


**An issue :(**
I did run into an issue related to authentication. So I have my production LXD instance running latest/stable and now a separate LXD instance running latest/edge. I went to the web ui of the `edge` variant, created a certificate and added it to my trust store in the latest/edge LXD. Added it to my laptop cert store and selected the cert to login to my latest/edge LXD. The usual process.

But now when I went back to my production LXD web ui to compare, I was logged out. I had to reselect a certificate to authenticate with. I selected my production lxd cert and logged back into the production LXD ui. But now I was logged out in the edge LXD web ui.

I can only be logged in to one of the LXD web UIs in a single browser (Edge). I went back and forth a few times before I gave up and opened Chrome to be able to be logged into both at the same time so that I can make a comparison.

This didn't happen before 4 weeks ago when I had two separate LXD instances with each their own separate certificates both running latest/stable.

I am using Microsoft Edge 130.0.2849.56  (latest) on Windows 11.

**Comment by edlerd** [comment]
Thanks for your kind feedback, let me reply to your questions:

> **The ability to export/import an instance.** i didn't expect it to immediately download the instance as a file to my laptop. Instead, I wanted to select a network filesystem on my LXD host. Maybe that is something that can be added in the future?

Allow me a question to clarify, because I am not sure if I am understanding your request correctly:
 - Do you expect a dialogue to select the target of the download file on your local host?
 - Or do you expect to store the export on the LXD server itself?
 
> **Migration** Great feature! Had to resort back to the cli two weeks ago to do this. Glad it's in the ui now, but I can only do one at a time. Can I request a feature to be able to select multiple VMs in the Instance overview so that we can migrate multiple instances at a time? I don't really want to have to do so many clicks for 50 separate instances.

Good suggestion, this is coming. We actually had bulk migration implemented, but then removed it due to performance concerns. When migrating many instances at once, the server might not be able to handle the requests in parallel. We have a backend story that is going to be implemented in the near future to add a queuing system for bulk actions. When that is done, the UI will also get bulk instance migrations.

> I did run into an issue related to authentication. So I have my production LXD instance running latest/stable and now a separate LXD instance running latest/edge. I went to the web ui of the latest/edge variant, created a certificate and added it to my trust store in the latest/edge LXD. Added it to my laptop cert store and selected the cert to login to my latest/edge LXD. The usual process. So far so good. [...]

You can reuse the same client certificate for multiple LXD servers. Maybe that already solves it for you? Unfortunately, browser support for client certificates is not very great, another option is to use a third party identity provider and rely on OIDC for authentication. You can find some preliminary [guidance on the OIDC setup in the lxd-ui wiki](https://github.com/canonical/lxd-ui/wiki/Setup-oidc-login). We plan to add this to the official documentation shortly as well.

**Comment by vosdev** [comment]
> Good suggestion, this is coming. We actually had bulk migration implemented, but then removed it due to performance concerns. When migrating many instances at once, the server might not be able to handle the requests in parallel. We have a backend story that is going to be implemented in the near future to add a queuing system for bulk actions. When that is done, the UI will also get bulk instance migrations.

That's great to hear! :)

> Do you expect a dialogue to select the target of the download file on your local host?
No, this would be a setting in my browser.
> Or do you expect to store the export on the LXD server itself?
I would like to have this as an option. Perhaps after clicking the Export button a pop-up can ask whether I would like to export it on the server or if I would like to download it. But I am not sure if this is achievable. LXD runs inside a snap and doesn't have access to the host's filesystems if I'm not mistaken


> You can reuse the same client certificate for multiple LXD servers. Maybe that already solves it for you? Unfortunately, browser support for client certificates is not very great, another option is to use a third party identity provider and rely on OIDC for authentication. You can find some preliminary [guidance on the OIDC setup in the lxd-ui wiki](https://github.com/canonical/lxd-ui/wiki/Setup-oidc-login). We plan to add this to the official documentation shortly as well.

Yeah I will probably have to resort to re-using the certificate. I was also thinking about storing it on a Yubi Key so that I can bring it with me.

I will see if I can find some time to play around with the OIDC setup. That would be a great addition to LXD! Like you say, certificates are limited in browsers. LXD was cli/api only and for that use-case certificates are a great and secure option, but now with the web UI coming alive OIDC/Oauth is a necessity.


I do have one more question. The [roadmap ](https://github.com/canonical/lxd-ui/blob/b33fd831671f83dc1c1d323a6df142fe68a07777/ROADMAP.md?plain=1#L16) for April 2024 mentions Ceph Management. What kind of Ceph Management is that referring to?


Thanks for the response :)

**Comment by edlerd** [comment]
> [...] after clicking the Export button a pop-up can ask whether I would like to export it on the server or if I would like to download it. But I am not sure if this is achievable. LXD runs inside a snap and therefore doesn't have access to the host's filesystems if I'm not mistaken

Yes, this is rather tricky, as we don't have file system access on the LXD server from the UI. I still don't understand the motivation. For backup purposes on the server, you might as well create an instance snapshot. Export is to transfer it to another server or to have an offline backup. So it seems counterintuitive to me to have the export placed on the LXD server.

> I do have one more question. The [roadmap ](https://github.com/canonical/lxd-ui/blob/b33fd831671f83dc1c1d323a6df142fe68a07777/ROADMAP.md?plain=1#L16) for April 2024 mentions Ceph Management. What kind of Ceph Management is that referring to?

This is referring to creating and configuring storage pools with the `ceph` driver. Good that you mention the roadmap, we just finished planning for 25.04, and I'll update the public one in the repository shortly.
